main/main.go

```diff
@@ -0,0 +1,275 @@
+		}
+	}
+	return cveEntries
+}
+
+func appliesToPrject(cve utils.NvdCVEEntry, project string) bool {
+	for _, vendorData := range cve.CVE.Affects.Vendor.VendorData {
+		for _, productData := range vendorData.Product.ProductData {
+			if productData.ProductName == project {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getDataFeeds(localPath string) (map[string]string, error) {
+	dataFeedHashes := make(map[string]string)
+	var dataFeedNames []string
+	for y := 2014; y <= time.Now().Year(); y++ {
+		dataFeedNames = append(dataFeedNames, strconv.Itoa(y))
+	}
+
+	// Get hashes for these feeds.
+	for _, dataFeedName := range dataFeedNames {
+		hash, err := getHashFromMetaURL(fmt.Sprintf(dataFeedMetaURL, dataFeedName))
+		if err != nil {
+			log.WithError(err).WithField(logDataFeedName, dataFeedName).Warning("could not get NVD data feed hash")
+			continue
+		}
+		dataFeedHashes[dataFeedName] = hash
+	}
+
+	// Create map containing the name and filename for every data feed.
+	dataFeedReaders := make(map[string]string)
+	for _, dataFeedName := range dataFeedNames {
+		fileName := filepath.Join(localPath, fmt.Sprintf("%s%s", dataFeedName, fileExt))
+		if _, ok := dataFeedHashes[dataFeedName]; ok {
+			// The hash is known, the disk should contains the feed. Try to read from it.
+			if localPath != "" {
+				if f, err := os.Open(fileName); err == nil {
+					f.Close()
+					dataFeedReaders[dataFeedName] = fileName
+					continue
+				}
+			}
+
+			err := downloadFeed(dataFeedName, fileName)
+			if err != nil {
+				return dataFeedReaders, err
+			}
+			dataFeedReaders[dataFeedName] = fileName
+		}
+	}
+	return dataFeedReaders, nil
+}
+
+func downloadFeed(dataFeedName, fileName string) error {
+	r, err := httputil.GetWithUserAgent(fmt.Sprintf(dataFeedURL, dataFeedName))
+	if err != nil {
+		log.WithError(err).WithField(logDataFeedName, dataFeedName).Error("could not download NVD data feed")
+		return commonerr.ErrCouldNotDownload
+	}
+	defer r.Body.Close()
+
+	if r.StatusCode != http.StatusOK {
+		log.WithFields(log.Fields{"StatusCode": r.StatusCode, "DataFeedName": dataFeedName}).Error("failed to download NVD data feed")
+		return commonerr.ErrCouldNotDownload
+	}
+
+	if !isGzipResponse(r) {
+		log.WithFields(log.Fields{"DataFeedName": dataFeedName}).Errorf("response is not a gzip encoded response")
+		return fmt.Errorf("response for datafeed: %q is not gzip encoded", dataFeedName)
+	}
+
+	err = utils.StoreHTTPResponseToFile(r, fileName, dataFeedName)
+	if err != nil {
+		os.Exit(1)
+	}
+	return nil
+}
+
+func isGzipResponse(r *http.Response) bool {
+	// Check Content-Encoding
+	for _, s := range r.Header["Content-Encoding"] {
+		if s == "gzip" {
+			return true
+		}
+	}
+	// If Content-Encoding is not set, check Content-Type
+	for _, s := range r.Header["Content-Type"] {
+		if strings.Contains(s, "gzip") {
+			return true
+		}
+	}
+	return false
+}
+
+func getHashFromMetaURL(metaURL string) (string, error) {
```

Comment:
Ah the hash value is used to match previous hash value, and if the hash is changed, then only new feed is downloaded.

Reply:
I don't think the code currently works that way; you'd need to store the hash in the GCS bucket as well to make this work.

Reply:
Since, we're running inside an ephemeral container, and writing files to a temp dir, I'm planning to get rid of the whole hash matching logic (where we check if current hash is same as previous hash and only download CVEs if hash are different). We're going to download to CVEs with every run of the job for every year from now on. Does that sound ok ?

Reply:
After discussion with @connorgorman , it is important that on our serving side we serve the hashes such that a central instance can now that it has the latest version of the data without having to download several megs. For the downloading side in the cronjob, it's fine to download everything from scratch. However, given that we need to somehow get the hashes to a GCS bucket, we *could* probably do it - just not a requirement.

Reply:
The data in this case is really small, since it's only Kube and Istio CVEs. It's going to be a few hundred KB at most. Unless I'm super off in my estimation.

Reply:
The k8sCVE file is 110K and istioCVE file is 7.6K. If it's still considerable amount of data, we can push the hash of the data of the files too from this repo.